doc/manual/src/language/syntax.md

```diff
@@ -183,6 +183,9 @@ This section covers syntax and semantics of the Nix language.
   Numbers, which can be *integers* (like `123`) or *floating point*
   (like `123.43` or `.27e13`).
 
+  Integers in the Nix language are 64-bit signed integers.
+  Integer overflow is defined to throw an error.
```

Comment:
why not just define it to be wrapping? It can be cheaply achieved by just telling the compiler `-fwrapv`. Throwing errors & extra checking the overflow always has significantly more overhead because you must check that for each computing.

Reply:
we benchmarked it. -ftrapv has bad overhead but using the sanitizer has *0 measurable overhead on either clang or gcc*. we don't *want* to allow signed overflow because the vast majority of the time it is a bug!

Reply:
`-fsanitize=signed-integer-overflow -fsanitize-undefined-trap-on-error` is what we're using

Reply:
> we benchmarked it. -ftrapv has bad overhead 

Where is the benchmark then?

>  0 measurable overhead on either clang or gcc. we don't want to allow signed overflow because the vast majority of the time it is a bug!

Why the overhead of this implementation has overhead comparable to sanitizers?

Reply:
Also please note that I mean `fwrapv`, not `ftrapv`.

Reply:
This code is different than the compiler flags which we *did* seriously benchmark when adding to lix. There's not enough arithmetic in the nix language to expect any difference from this PR, but I can drag out the benchmark scripts and confirm it.

Reply:
> Where is the benchmark of -fwrapv then?

That is a non-sequitur to this change, which is about correctness of the Nix language. Separately I think enabling `-fwrapv` globally is a terribly bad idea because it allows overflow bugs in unrelated parts of CppNix to become silent data corruptions instead of the alternative of turning them into crashes with no visible performance impact to our use case.

Here is us doing the signed overflow to crash thing: https://gerrit.lix.systems/c/lix/+/626

I don't know if the benchmark results from `-ftrapv` survived, but they were notable on gcc. We never benchmarked `-fwrapv` since it would have *at best* worse performance than letting it remain UB, and we could get a correctness improvement by turning it into a crash for essentially free.

> Why the overhead of this implementation has overhead comparable to sanitizers?

This implementation is unrelated to the idea of globally banning overflow (which we did months ago in Lix) and only affects the Nix language, and based on the benchmarks below, *this change* has insignificant performance impact of a magnitude that could equally be expected to be caused by linking order changing or other such things. In return, it fixes both C++ and Nix language-level undefined behaviour by using arithmetic with fully defined semantics in C++ to produce fully defined Nix language semantics.